pkg/discovery/processor/cluster_processor.go

```diff
@@ -65,68 +71,107 @@ func (processor *ClusterProcessor) ConnectCluster() error {
 	return nil
 }
 
+// Perform a single node validation
+func (processor *ClusterProcessor) checkNodesWorker(work chan *v1.Node, done chan bool, index int) {
+	glog.Infof("node verifier worker %d starting", index)
+	for {
+		node, present := <-work
+		if !present {
+			glog.Infof("node verifier worker %d finished. No more work", index)
+			return
+		}
+		nodeCpuFrequency, err := checkNode(node, processor.nodeScrapper)
+		if err != nil {
+			glog.Errorf("%s [err:%s]", node.Name, err)
+		} else {
+			// Log the success and send the response to everybody
+			glog.Infof("verified %s [cpu:%v MHz]", node.Name, nodeCpuFrequency)
+			done <- true
+			// Force return here. We are done and notified everybody.
+			glog.Infof("node verifier worker %d finished. Successful verification", index)
+			return
+		}
+	}
+}
+
+// Wait for at least one of the workers to complete successfully
+// or timeout
+func waitForCompletion(done chan bool) bool {
+	cycles := int(totalWaitTime.Nanoseconds() / sleepInterval.Nanoseconds())
+	for i := 0; i < cycles; i++ {
+		select {
+		case <-done:
+			return true
+		default:
+			time.Sleep(sleepInterval)
+		}
+	}
+	return false
+}
+
+// Drains the work queue.
+// This will allow us to terminate the still on worker threads
+func drainWorkQueue(work chan *v1.Node) {
+	for {
+		_, ok := <-work
+		if !ok {
+			return
+		}
+	}
+}
+
 // Connects to all nodes in the cluster.
 // Obtains the CPU Frequency of the node to determine the node availability and accessibility.
 func (processor *ClusterProcessor) connectToNodes() (*ClusterValidationResult, error) {
 	var unreachable []string
 	validationResult := &ClusterValidationResult{
 		UnreachableNodes: unreachable,
 	}
-
 	nodeList, err := processor.clusterInfoScraper.GetAllNodes()
 	if err != nil {
 		validationResult.IsValidated = false
 		return validationResult, fmt.Errorf("Error getting nodes for cluster : %s\n", err)
 	}
 	glog.V(2).Infof("There are %d nodes\n", len(nodeList))
-
-	var connectionErrors []string
-	var connectionMsgs []string
-
-	for _, node := range nodeList {
-		nodeCpuFrequency, err := checkNode(node, processor.nodeScrapper)
-		if err != nil {
-			connectionErrors = append(connectionErrors, fmt.Sprintf("%s [err:%s]", node.Name, err))
-			unreachable = append(unreachable, node.Name)
-		} else {
-			connectionMsgs = append(connectionMsgs,
-				fmt.Sprintf("%s [cpu:%v MHz]", node.Name, nodeCpuFrequency))
-		}
+	// The connection data
+	size := len(nodeList)
+	work := make(chan *v1.Node, size)
```

Comment:
I think we can optimize this `single-producer, multi-consumers` pattern by better naming fo data input/output channels, control channels.